### PR TITLE
Adds pet carriers

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: You can now make pet carriers from the autolathe, to carry around chef meat and other small animals without having to drag them. The HoP, captain, and CMO also start with carriers in their lockers for their pets.
/:cl:

**Sprites by WalterJe from his forum [thread](https://tgstation13.org/phpBB/viewtopic.php?f=11&t=1659), thank you!**

Have you ever dragged Ian around the station, only for a random tider to steal them away? Now you can still have your pet stolen by anyone with a bottle of water, but now in *style!* This adds pet carriers to several head of staff lockers and to the Misc section of the autolathe.

You can attack any mob with a pet carrier to attempt to load it into the pet carrier. This only works by default for mobs of `MOB_SIZE_SMALL` size or below, like cats, dogs, and mice. It can hold one of these, or three tiny mobs like butterflies.

The carrier has a lock switch that prevents anything from pushing out. Sentient animals can resist out by moving, which will cause them to reach out and flip the lock switch after a delay between 30 and 40 seconds. (That's not realistic, but it prevents drones and sentient pets from being trapped forever.)

Admins can also edit variables to allow the carriers to carry any number of mobs and any weight size, meaning you can TC-trade for a ghetto version of His Grace.

I added this since I felt like it made more sense than dragging pets around the entire station manually, and also because it opens up creative things like putting viscerators and bees in a carrier.

### CODE REVIEW

**Comment by Repukan** [comment]
Make it able to carry manlets from genetics

**Comment by FantasticFwoosh** [comment]
We need a bluespace version that will allow us to shove catpeople & lizardpeople inside like the dirty animals they are.

Aside like this, this sounds like it could be fun for xenobiology for collecting stashes of gold-mobs.

**Comment by Xhuis** [comment]
"Amanita, I choose you!"
*Urist McScience has thrown the pet carrier.*
*Ian has been hit by the pet carrier.*
*Ian stops moving...*

**Comment by coiax** [comment]
Does moving while holding the pet carrier reset the resist timer?

**Comment by Xhuis** [comment]
It doesn't; the timer checks the mob itself, not the carrier, meaning unless the mob is interrupted somehow (i.e. taking another action, or if the owner lets it out without it resisting) then it'll break free regardless.

**Comment by coiax** [comment]
Are you sure though? Because you can be stuck on someone as a drone if they keep moving, because their movement interrupts your struggle attempt.

**Comment by Xhuis** [comment]
In drone code:
```
/obj/item/clothing/head/drone_holder/proc/uncurl()
	if(!drone)
		return

	if(isliving(loc))
		var/mob/living/L = loc
		to_chat(L, "<span class='warning'>[drone] is trying to escape!</span>")
		if(!do_after(drone, 50, target = L))
			return
		L.dropItemToGround(src)
```
`do_after` includes the wearer as its target. If the wearer moves at all, it cancels. Now, look in pet carrier code:
```
/obj/item/pet_carrier/container_resist(mob/living/user)
 	user.changeNext_move(CLICK_CD_BREAKOUT)
 	user.last_special = world.time + CLICK_CD_BREAKOUT
 	if(user.mob_size <= MOB_SIZE_SMALL)
 		to_chat(user, "<span class='notice'>You poke a limb through [src]'s bars and start fumbling for the lock switch... (This will take some time.)</span>")
 		to_chat(loc, "<span class='warning'>You see [user] reach through the bars and fumble for the lock switch!</span>")
 		if(!do_after(user, rand(300, 400), target = user) || open || !locked || !user in occupants)
 			return
 		loc.visible_message("<span class='warning'>[user] flips the lock switch on [src] by reaching through!</span>", ignored_mob = user)
 		to_chat(user, "<span class='boldannounce'>Bingo! The lock pops open!</span>")
 		locked = FALSE
 		playsound(src, 'sound/machines/boltsup.ogg', 30, TRUE)
 		update_icon()
```
`do_after` includes the animal itself as a target, meaning that it checks for the animal itself, not anything to do with the carrier or its holder. This was an issue back in shadowling, too, when it was possible to enthrall someone at range because if someone put you in a locker and dragged you away it didn't technically interrupt you. So yes, I'm sure!

**Comment by FantasticFwoosh** [comment]
@Xhuis throw in feeding walking mushrooms evolutionary trigger reagents in candy sweets and its a finished feature if you want to walk around with your champion walking mushroom electric type in a pet carrier everywhere.

Eitherway the pet carrier will be good for holding/capturing chickens and misc livestock, hmm i wonder if we could replace the big wooden boxes that transport small animals in cargo?

**Comment by Bawhoppen** [comment]
Don't add any of that stuff Fwoosh just said

**Comment by FantasticFwoosh** [comment]
@Xhuis @You to @You 



> dont add any of that stuff fwoosh just said

>> **add it**

File: code/modules/research/designs/autolathe_designs.dm [hunk 1]
```diff
@@ -799,3 +799,11 @@
...
 	build_path = /obj/item/pet_carrier
 	category = list("initial", "Misc")
+
+/datum/design/miniature_power_cell
+	name = "Light Fixture Battery"
+	id = "miniature_power_cell"
+	build_type = AUTOLATHE
+	materials = list(MAT_GLASS = 20)
+	build_path = /obj/item/stock_parts/cell/emergency_light
+	category = list("initial", "Electronics")
```
Comment (coiax) [inline code/modules/research/designs/autolathe_designs.dm]: wait isn't this from your red light PR?

File: code/modules/research/designs/autolathe_designs.dm [hunk 2]
```diff
@@ -799,3 +799,11 @@
 	materials = list(MAT_METAL = 7500, MAT_GLASS = 100)
 	build_path = /obj/item/pet_carrier
 	category = list("initial", "Misc")
+
+/datum/design/miniature_power_cell
+	name = "Light Fixture Battery"
```
Comment (coiax) [inline code/modules/research/designs/autolathe_designs.dm]: also, item names should be `light fixture battery`

File: code/modules/research/designs/autolathe_designs.dm (hunk 1 above)
Comment (Xhuis) [reply code/modules/research/designs/autolathe_designs.dm]: You're looking at the commit that fixed conflicts!

File: code/modules/research/designs/autolathe_designs.dm (hunk 2 above)
Comment (Xhuis) [reply code/modules/research/designs/autolathe_designs.dm]: Autolathe designs have always been capitalized.

**Comment by BeeSting12** [comment]
can i put catgirls in them nyaa

**Comment by Xhuis** [comment]
I did include an amusing error message if you tried.
```
if(iscatperson(H))
     to_chat(user, "<span class='warning'>You'd need a lot of catnip and treats, plus maybe a laser pointer, for that to work.</span>")
```

**Comment by FantasticFwoosh** [comment]
Have the laser pointer in the offhand & make the dream real.

You know people will try to immediately, if not catpeople will try stuffing themselves into pet boxes just to **_RP_** and be obnoxious.

File: code/game/objects/items/pet_carrier.dm [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	var/open = TRUE
+	var/locked = FALSE
+	var/list/occupants = list()
+	var/occupant_weight = 0
+	var/max_occupants = 3 //Hard-cap so you can't have infinite mice or something in one carrier
+	var/max_occupant_weight = MOB_SIZE_SMALL //This is calculated from the mob sizes of occupants
+
+/obj/item/pet_carrier/Initialize()
+	. = ..()
+	START_PROCESSING(SSfastprocess, src)
```
Comment (optimumtact) [inline code/game/objects/items/pet_carrier.dm]: really, is this seriously what you're going to do to track occupants

File: code/game/objects/items/pet_carrier.dm [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+/obj/item/pet_carrier/Destroy()
+	if(occupants.len)
+		for(var/V in occupants)
+			remove_occupant(V)
+	STOP_PROCESSING(SSfastprocess, src)
+	return ..()
+
+/obj/item/pet_carrier/process()
```
Comment (optimumtact) [inline code/game/objects/items/pet_carrier.dm]: this is garbage code

**Comment by coiax** [comment]
> Export live facehuggers or just keep em contained inside, sure whatever.

Facehuggers aren't mobs, they're items, specifically clothing masks.

File: code/game/objects/items/pet_carrier.dm (hunk 3 above)
Comment (Xhuis) [reply code/game/objects/items/pet_carrier.dm]: What would you rather I do? Check for nulls on removal and adding?

**Comment by FantasticFwoosh** [comment]
@coiax true but you can just add them onto a allowed list of contents if you're fast enough to grab one without it attempting to leap on you. Slightly neater than putting it inside a locker.

File: code/game/objects/items/pet_carrier.dm (hunk 3 above)
Comment (AnturK) [reply code/game/objects/items/pet_carrier.dm]: Just use Exited. This absolute does not need to process anything.

Comment (Xhuis) [reply code/game/objects/items/pet_carrier.dm]: I wrote this code at like 1 AM, cut me some slack.

**Comment by coiax** [comment]
If a facehugger doesn't see you as a valid target (you're pregnant or you have an alien hive node), then you can just pick it up and put it in your backpack.

**Comment by FantasticFwoosh** [comment]
Last time i checked picking up facehuggers smushed them, but eitherway i guess having that node &/or relevant protections would make transporting them to places like cargo bay from your 'oh god why' xenomorph corral immediately safer.

Letting us do more stuff economically such as attaching a export price to them.

**Comment by imsxz** [comment]
They only get smashed if you're able to get preggers but wearing a protective suit when you try to pick it up. They smash themselves on your helmet/whatever else you might be wearing on your head.

File: code/game/objects/items/pet_carrier.dm [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	if(occupant in occupants && isliving(occupant))
+		var/mob/living/L = occupant
+		occupants -= occupant
+		occpuant_weight -= L.mob_size
+
+/obj/item/pet_carrier/handle_atom_del(atom/A)
+	if(A in occupants && isliving(A))
+		var/mob/living/L = occupant
+		occupants -= L
+		occpuant_weight -= L.mob_size
```
Comment (optimumtact) [inline code/game/objects/items/pet_carrier.dm]: isn't this occupant_weight

Comment (Xhuis) [reply code/game/objects/items/pet_carrier.dm]: Frick, it even compiled before I changed the *one* line and I thought it was fine ugh.

File: code/game/objects/items/pet_carrier.dm [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		to_chat(user, "<span class='warning'>[src] is too heavy to hold any more animals!</span>")
+		return
+	user.visible_message("<span class='notice'>[user] starts loading [target] into [src].</span>", \
+	"<span class='notice'>You start loading [target] into [src]...</span>", ignored_mob = target)
+	to_chat(target, "<span class='userdanger'>[user] starts loading you into their [name]!</span>")
+	if(!do_mob(user, target, 30))
+		return
+	if(target in occupants)
+		return
+	if(occupants.len >= max_occupants) //Run the checks again, just in case
```
Comment (optimumtact) [inline code/game/objects/items/pet_carrier.dm]: pop these duplicate checks in a proc and just call it before and after

Comment (optimumtact) [reply code/game/objects/items/pet_carrier.dm]: or a macro

File: code/modules/research/designs/autolathe_designs.dm (hunk 2 above)
Comment (coiax) [reply code/modules/research/designs/autolathe_designs.dm]: my mistake

File: code/game/objects/items/pet_carrier.dm [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	else
+		if(locked)
+			to_chat(user, "<span class='warning'>[src] is locked!</span>")
+			return
+		to_chat(user, "<span class='notice'>You open [src]'s door.</span>")
+		playsound(user, 'sound/effects/bin_open.ogg', 50, TRUE)
+		open = TRUE
+	update_icon()
+
+/obj/item/pet_carrier/AltClick(mob/living/user)
```
Comment (AnturK) [inline code/game/objects/items/pet_carrier.dm]: Ghost alt click

**Review by AnturK — COMMENTED** [review]
Bit late but

File: code/game/objects/items/pet_carrier.dm (hunk 7 above)
Comment (vuonojenmustaturska) [reply code/game/objects/items/pet_carrier.dm]: `if (open || !user.canUseTopic(src, be_close=TRUE))`

Comment (vuonojenmustaturska) [reply code/game/objects/items/pet_carrier.dm]: oh wait this got merged

**Pull request merged**